### ISSUE TITLE
Project edit page becomes inaccessible; api returns 500 errors; null pointer exception in logs

### ISSUE THREAD

**Issue by cjellick** [rancher/rancher#10056]
rancher/server:v1.6.5

A user experienced a problem where after they delete a user from their ldap (They are using *FreeIPA*), the project edit page became inaccessible for projects where the delete user was a member because the API was returning a 500 error. Corresponding to that 500 was the following error in the logs:
```
2017-10-04 20:36:03,266 ERROR [:] [] [] [] [1020391880-2519] [i.g.i.g.r.handler.ExceptionHandler  ] Exception in API for request [https://rancher.homeadvisor.com/v2-beta/projects/1a7/projectmembers]. Error id: [87ec2d8a-4219-4beb-9e44-3ac735913809]. java.lang.NullPointerException: null
        at io.cattle.platform.iaas.api.auth.integration.ldap.LDAPIdentityProvider.isType(LDAPIdentityProvider.java:240) ~[cattle-iaas-auth-logic-0.5.0-SNAPSHOT.jar:na]
        at io.cattle.platform.iaas.api.auth.integration.ldap.LDAPIdentityProvider.getObject(LDAPIdentityProvider.java:168) ~[cattle-iaas-auth-logic-0.5.0-SNAPSHOT.jar:na]
        at io.cattle.platform.iaas.api.auth.integration.ldap.LDAPIdentityProvider.getIdentity(LDAPIdentityProvider.java:79) ~[cattle-iaas-auth-logic-0.5.0-SNAPSHOT.jar:na]
        at io.cattle.platform.iaas.api.auth.identity.IdentityManager.projectMemberToIdentity(IdentityManager.java:139) ~[cattle-iaas-auth-logic-0.5.0-SNAPSHOT.jar:na]
        at io.cattle.platform.iaas.api.auth.identity.IdentityManager.projectMemberToIdentity(IdentityManager.java:169) ~[cattle-iaas-auth-logic-0.5.0-SNAPSHOT.jar:na]
        at io.cattle.platform.iaas.api.auth.projects.ProjectMemberResourceManager.listInternal(ProjectMemberResourceManager.java:86) ~[cattle-iaas-auth-logic-0.5.0-SNAPSHOT.jar:na]
        at io.github.ibuildthecloud.gdapi.request.resource.impl.AbstractBaseResourceManager.list(AbstractBaseResourceManager.java:82) ~[cattle-framework-java-server-0.5.0-SNAPSHOT.jar:na]
        at io.cattle.platform.api.resource.AbstractObjectResourceManager.getChildLink(AbstractObjectResourceManager.java:250) ~[cattle-framework-api-0.5.0-SNAPSHOT.jar:na]
        at io.cattle.platform.api.resource.AbstractObjectResourceManager.getLinkInternal(AbstractObjectResourceManager.java:208) ~[cattle-framework-api-0.5.0-SNAPSHOT.jar:na]
        at io.github.ibuildthecloud.gdapi.request.resource.impl.AbstractBaseResourceManager.getLink(AbstractBaseResourceManager.java:146) ~[cattle-framework-java-server-0.5.0-SNAPSHOT.jar:na]
        at io.github.ibuildthecloud.gdapi.request.resource.AbstractResourceManagerFilter.getLink(AbstractResourceManagerFilter.java:28) ~[cattle-framework-java-server-0.5.0-SNAPSHOT.jar:na]
        at io.github.ibuildthecloud.gdapi.request.resource.impl.FilteredResourceManager.getLink(FilteredResourceManager.java:41) ~[cattle-framework-java-server-0.5.0-SNAPSHOT.jar:na]
        at io.github.ibuildthecloud.gdapi.request.handler.ResourceManagerRequestHandler.generate(ResourceManagerRequestHandler.java:52) ~[cattle-framework-java-server-0.5.0-SNAPSHOT.jar:na]
        at io.github.ibuildthecloud.gdapi.request.handler.AbstractResponseGenerator.handle(AbstractResponseGenerator.java:14) ~[cattle-framework-java-server-0.5.0-SNAPSHOT.jar:na]
        at io.github.ibuildthecloud.gdapi.request.handler.write.DefaultReadWriteApiDelegate.handle(DefaultReadWriteApiDelegate.java:27) ~[cattle-framework-java-server-0.5.0-SNAPSHOT.jar:na]
        at io.github.ibuildthecloud.gdapi.request.handler.write.DefaultReadWriteApiDelegate.read(DefaultReadWriteApiDelegate.java:17) ~[cattle-framework-java-server-0.5.0-SNAPSHOT.jar:na]
        at sun.reflect.GeneratedMethodAccessor444.invoke(Unknown Source) ~[na:na]
        ...
```

I setup FreeIPA successfully with Rancher, but I've not been able to reproduce the 500. However, the stack trace makes the problem fairly obvious and easy to fix.

**Comment by cjellick** [comment]
Can be tested on v1.6-development as of time/date of this comment

**Comment by cjellick** [comment]
@sangeethah I'm assigning to you but feel free to re-assign to a different QA

Whoever picks this up for testing, let's chat. TL;DR I haven't been able to reproduce, so we'll have to get creative for testing.

**Comment by cjellick** [comment]
Random notes:
- Good resource on getting FreeIPA up and running: https://github.com/freeipa/freeipa-workshop#unit-1-installing-the-freeipa-server
- To attempt to repro, one lead I followed up on was doing --preserve when deleting the user: http://www.freeipa.org/page/V4/User_Life-Cycle_Management#user_plugin
- If using the ipa cli, dont forget to start with `kinit` to get your creds in there https://www.redhat.com/archives/freeipa-users/2014-April/msg00113.html
- A few random links I had open in the browser at the time of writing this related to configuring/using FreeIPA:
  - https://access.redhat.com/documentation/en-US/Red_Hat_Satellite/6.1/html/User_Guide/chap-Red_Hat_Satellite-User_Guide-Configuring_External_Authentication.html
  - https://www.freeipa.org/page/Directory_Server
  - https://www.freeipa.org/page/HowTo/LDAP

**Comment by cjellick** [comment]
Also, here's the config that worked for me when configuring FreeIPA:
```
mysql> select * from setting;
+----+-------------------------------------------------------+------------------------------------------------------------------+
| id | name                                                  | value                                                            |
+----+-------------------------------------------------------+------------------------------------------------------------------+
|  1 | api.auth.ldap.openldap.group.dn.field                 | entryDN                                                          |
|  2 | api.auth.ldap.openldap.group.member.user.attribute    | entryDN                                                          |
|  3 | install.uuid                                          | 3a276208-c338-40ca-9c20-b733478abe26                             |
|  4 | api.auth.ldap.openldap.access.mode                    | unrestricted                                                     |
|  5 | api.auth.ldap.openldap.domain                         | cn=users,cn=accounts,dc=ipademo,dc=local                         |
|  6 | api.auth.ldap.openldap.group.search.domain            |                                                                  |
|  7 | api.auth.ldap.openldap.group.name.field               | cn                                                               |
|  8 | api.auth.ldap.openldap.group.object.class             | posixGroup                                                       |
|  9 | api.auth.ldap.openldap.group.search.field             | cn                                                               |
| 10 | api.auth.ldap.openldap.group.member.mapping.attribute | memberUid                                                        |
| 11 | api.auth.ldap.openldap.login.domain                   |                                                                  |
| 12 | api.auth.ldap.openldap.port                           | 389                                                              |
| 13 | api.auth.ldap.openldap.server                         | server.ipademo.local                                             |
| 14 | api.auth.ldap.openldap.service.account.password       | *****                                                        |
| 15 | api.auth.ldap.openldap.service.account.user           | uid=admin,cn=users,cn=accounts,dc=ipademo,dc=local               |
| 16 | api.auth.ldap.openldap.tls                            | false                                                            |
| 17 | api.auth.ldap.openldap.user.enabled.mask.bit          | 0                                                                |
| 18 | api.auth.ldap.openldap.user.enabled.attribute         |                                                                  |
| 19 | api.auth.ldap.openldap.user.login.field               | uid                                                              |
| 20 | api.auth.ldap.openldap.user.member.attribute          | memberOf                                                         |
| 21 | api.auth.ldap.openldap.user.name.field                | cn                                                               |
| 22 | api.auth.ldap.openldap.user.object.class              | inetuser                                                         |
| 23 | api.auth.ldap.openldap.user.search.field              | uid                                                              |
| 24 | api.auth.ldap.openldap.connection.timeout             | 5000                                                             |
| 25 | api.security.enabled                                  | true                                                             |
| 26 | api.auth.provider.configured                          | openldapconfig                                                   |
| 27 | api.auth.ldap.openldap.allowed.identities             |                                                                  |
| 28 | api.auth.enabler                                      | openldap_user:uid=admin,cn=users,cn=accounts,dc=ipademo,dc=local |
```

**Comment by galal-hussein** [comment]
Rancher server: v1.6.11-rc3

After installation FreeIPA and authenticating with Rancher, here are my steps:

- Create new environments and add FreeIPA users.
- I deleted these users using the Web UI and CLI with and without --preserve option.
- Try to edit environments that have the deleted users.

I was able to edit the environments normally without errors, i can't reproduce the mentioned problem.

**Comment by mwelch-ptc** [comment]
I realize this is an old closed issue but in both this issue and #9380, the proposed workaround was to re-add the user to LDAP, then delete them from the UI. Well, that works when you have control of LDAP but in most corporate environments, like ours, our LDAP group is NOT going to add a former employee back to LDAP for us. Nope. 

Any other methods of getting around this?
